File: app/servicos/anexos.py
"""RN-12 - anexos com dedup por SHA-256 e blob compartilhado."""

from __future__ import annotations

import hashlib
import shutil
from dataclasses import dataclass
from pathlib import Path

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.config import obter_config
from app.modelos import Anexo
from app.servicos import auditoria
from app.servicos.rbac import UsuarioAtual
from app.servicos.textos import slug_ascii

# O que se anexa a um processo do adicional ocupacional. E o que a tela do
# processo oferece: colocar ali a ficha de EPI ou a CAT/SP seria oferecer a
# categoria errada no lugar errado.
CATEGORIAS_PROCESSO = (
    "PARECER_ASSINADO",
    "LAUDO",
    "PORTARIA",
    "DESPACHO",
    "FORMULARIO",
    "RELATORIO_CAMPO",
    "FOTO",
    "OUTRO",
)

# Categorias dos modulos que ainda estao sendo construidos. Entraram todas de
# uma vez, antes das telas que as usam, porque estender `ck_anexo_cat` no SQLite
# reconstroi a tabela `anexo` — que ja tem dados e e referenciada por FK. Uma
# reconstrucao em vez de tres.
#
# Categoria nao e rotulo: e ela que decide QUANDO o arquivo pode ser eliminado.
# `docs/POLITICA_RETENCAO.md` §2.1 da 5 anos a RELATORIO_CAMPO/FOTO/OUTRO e
# guarda permanente a PARECER_ASSINADO/LAUDO/PORTARIA/FORMULARIO. Gravar a
# rubrica do instrutor como 'FOTO' seria marcar para eliminacao em 5 anos a
# assinatura de um certificado de guarda permanente — daí RUBRICA_INSTRUTOR ter
# nome proprio. Nenhuma das oito novas esta nomeada naquelas duas linhas ainda:
# hoje isso nao apaga nada (nao existe rotina de eliminacao no sistema), mas a
# fatia que criar a rotina precisa das linhas antes.
CATEGORIAS_EPI = ("MANUAL_EPI", "FICHA_EPI")
CATEGORIAS_CERTIFICADOS = ("RUBRICA_INSTRUTOR", "LISTA_PRESENCA", "CERTIFICADO")
CATEGORIAS_ACIDENTES = ("CAT_SP", "RELATORIO_INVESTIGACAO", "EVIDENCIA_ACIDENTE")

# O conjunto que `guardar` aceita e que `ck_anexo_cat` admite — as duas listas
# tem de casar, e um teste compara.
CATEGORIAS = (
    CATEGORIAS_PROCESSO + CATEGORIAS_EPI + CATEGORIAS_CERTIFICADOS + CATEGORIAS_ACIDENTES
)

# Categoria que nasce restrita. O criterio e o mesmo desde a RN-21: o documento
# nomeia pessoa e diz algo sobre a condicao dela no trabalho.
#   FORMULARIO      - o do art. 17 traz declaracao de exposicao
#   FICHA_EPI       - prova de entrega, nominal (IN 15/2022 e NR-6)
#   LISTA_PRESENCA  - assinatura de quem esteve na turma
#   CERTIFICADO     - o PDF nominal; a validacao publica e por chave, nao pelo arquivo
#   CAT_SP e os dois de acidente - evento com a pessoa identificada
# MANUAL_EPI e RUBRICA_INSTRUTOR ficam publicos: manual de fabricante nao tem
# titular, e a rubrica ja sai impressa em todo certificado.
NIVEL_POR_CATEGORIA = {
    "FORMULARIO": "RESTRITO",
    "FICHA_EPI": "RESTRITO",
    "LISTA_PRESENCA": "RESTRITO",
    "CERTIFICADO": "RESTRITO",
    "CAT_SP": "RESTRITO",
    "RELATORIO_INVESTIGACAO": "RESTRITO",
    "EVIDENCIA_ACIDENTE": "RESTRITO",
}


@dataclass
class ResultadoAnexo:
    anexo: Anexo
    duplicado: bool
    tambem_em: int


def digerir(conteudo: bytes) -> str:
    return hashlib.sha256(conteudo).hexdigest()


def caminho_do_blob(sha256: str) -> Path:
    cfg = obter_config()
    base = cfg.caminho(cfg.dir_anexos) / sha256[:2]
    base.mkdir(parents=True, exist_ok=True)
    return base / sha256


def guardar(
    s: Session,
    *,
    entidade: str,
    entidade_id: int,
    nome_original: str,
    conteudo: bytes,
    mime_type: str,
    categoria: str,
    usuario: UsuarioAtual,
    processo_id: int | None = None,
    numero_documento_sei: str | None = None,
    assinado: bool = False,
    origem_migracao: str | None = None,
    origem_ref: str | None = None,
) -> ResultadoAnexo:
    if categoria not in CATEGORIAS:
        raise ValueError(f"categoria inválida: {categoria}")

    sha256 = digerir(conteudo)

    existente = s.execute(
        select(Anexo).where(
            Anexo.entidade == entidade,
            Anexo.entidade_id == entidade_id,
            Anexo.sha256 == sha256,
        )
    ).scalar_one_or_none()
    if existente is not None:
        auditoria.registrar(
            s,
            entidade="anexo",
            entidade_id=existente.id,
            processo_id=processo_id,
            tipo_evento=auditoria.ANEXO_DUPLICADO,
            descricao=(
                f"Arquivo '{nome_original}' ignorado: já anexado como "
                f"'{existente.nome_original}' (mesmo SHA-256)."
            ),
            usuario=usuario,
        )
        return ResultadoAnexo(existente, True, _quantos_processos(s, sha256))

    if categoria == "PARECER_ASSINADO":
        anterior = s.execute(
            select(Anexo).where(
                Anexo.entidade == entidade,
                Anexo.entidade_id == entidade_id,
                Anexo.categoria == "PARECER_ASSINADO",
                Anexo.ativo.is_(True),
            )
        ).scalar_one_or_none()
        if anterior is not None:
            anterior.ativo = False
            s.flush()

    destino = caminho_do_blob(sha256)
    if not destino.exists():
        destino.write_bytes(conteudo)

    anexo = Anexo(
        entidade=entidade,
        entidade_id=entidade_id,
        nome_arquivo=slug_ascii(Path(nome_original).stem)[:200]
        + Path(nome_original).suffix.lower(),
        nome_original=nome_original,
        mime_type=mime_type or "application/octet-stream",
        tamanho_bytes=len(conteudo),
        sha256=sha256,
        storage_key=str(destino.relative_to(obter_config().caminho(obter_config().dir_anexos))),
        categoria=categoria,
        nivel_acesso=NIVEL_POR_CATEGORIA.get(categoria, "PUBLICO"),
        numero_documento_sei=numero_documento_sei,
        assinado=assinado,
        origem_migracao=origem_migracao,
        origem_ref=origem_ref,
        enviado_por=usuario.id,
    )
    s.add(anexo)
    s.flush()

    auditoria.registrar(
        s,
        entidade="anexo",
        entidade_id=anexo.id,
        processo_id=processo_id,
        tipo_evento="ANEXO_ENVIADO",
        descricao=f"'{nome_original}' ({categoria}, {len(conteudo)} bytes).",
        usuario=usuario,
    )
    return ResultadoAnexo(anexo, False, _quantos_processos(s, sha256))


def guardar_arquivo(
    s: Session, caminho: Path, **kwargs
) -> ResultadoAnexo:
    return guardar(s, conteudo=caminho.read_bytes(), nome_original=caminho.name, **kwargs)


def _quantos_processos(s: Session, sha256: str) -> int:
    linhas = s.execute(
        select(Anexo.entidade, Anexo.entidade_id).where(Anexo.sha256 == sha256)
    ).all()
    return len({(e, i) for e, i in linhas})


def listar(s: Session, entidade: str, entidade_id: int) -> list[Anexo]:
    return list(
        s.execute(
            select(Anexo)
            .where(
                Anexo.entidade == entidade,
                Anexo.entidade_id == entidade_id,
                Anexo.ativo.is_(True),
            )
            .order_by(Anexo.enviado_em.desc())
        ).scalars()
    )


def caminho_absoluto(anexo: Anexo) -> Path:
    cfg = obter_config()
    return cfg.caminho(cfg.dir_anexos) / anexo.storage_key


def copiar_para(anexo: Anexo, destino: Path) -> Path:
    destino.parent.mkdir(parents=True, exist_ok=True)
    shutil.copy2(caminho_absoluto(anexo), destino)
    return destino

File: nuvem/src/servicos/anexos.ts
/**
 * RN-12 — anexos com dedup por SHA-256 e blob compartilhado.
 * Porte de `app/servicos/anexos.py`.
 *
 * O blob é endereçado pelo conteúdo (`<sha[:2]>/<sha>`, o mesmo
 * `storage_key` do Python) e mora no armazenamento (`armazenamento.ts`:
 * Supabase Storage na nuvem, pasta local em desenvolvimento), sob o prefixo
 * `anexos/`. O mesmo arquivo anexado a dois processos ocupa um blob só.
 *
 * **Desvio: limite de tamanho** (ver DESVIOS.md). O Python não tinha limite —
 * o arquivo chegava pela rede local do setor. Na nuvem o corpo da requisição
 * da função do Netlify para em 6 MB, e o binário do multipart chega à função
 * codificado (≈4,5 MB úteis). Arquivo acima de `LIMITE_BYTES` é recusado aqui
 * com a saída escrita, em vez de estourar como 500 sem explicação.
 */
import { createHash } from "node:crypto";
import { and, desc, eq } from "drizzle-orm";
import type { Executor } from "../db/cliente.js";
import { anexo as tabela_anexo } from "../db/esquema/index.js";
import { RegraViolada } from "../nucleo/erros.js";
import * as auditoria from "./auditoria.js";
import { obterArmazenamento } from "./armazenamento.js";
import type { UsuarioAtual } from "./rbac.js";
import { slug_ascii } from "./textos.js";

export type Anexo = typeof tabela_anexo.$inferSelect;

// O que se anexa a um processo do adicional ocupacional. É o que a tela do
// processo oferece.
export const CATEGORIAS_PROCESSO = [
  "PARECER_ASSINADO",
  "LAUDO",
  "PORTARIA",
  "DESPACHO",
  "FORMULARIO",
  "RELATORIO_CAMPO",
  "FOTO",
  "OUTRO",
] as const;

// Categoria não é rótulo: é ela que decide QUANDO o arquivo pode ser
// eliminado (`docs/POLITICA_RETENCAO.md` §2.1) — daí RUBRICA_INSTRUTOR ter
// nome próprio em vez de 'FOTO'.
export const CATEGORIAS_EPI = ["MANUAL_EPI", "FICHA_EPI"] as const;
export const CATEGORIAS_CERTIFICADOS = ["RUBRICA_INSTRUTOR", "LISTA_PRESENCA", "CERTIFICADO"] as const;
export const CATEGORIAS_ACIDENTES = ["CAT_SP", "RELATORIO_INVESTIGACAO", "EVIDENCIA_ACIDENTE"] as const;

// O conjunto que `guardar` aceita e que `ck_anexo_cat` admite — as duas listas
// têm de casar, e um teste compara. A lista do CHECK é a de `dominio/auditoria`.
export const CATEGORIAS: readonly string[] = [
  ...CATEGORIAS_PROCESSO,
  ...CATEGORIAS_EPI,
  ...CATEGORIAS_CERTIFICADOS,
  ...CATEGORIAS_ACIDENTES,
];

// Categoria que nasce restrita: o documento nomeia pessoa e diz algo sobre a
// condição dela no trabalho.
export const NIVEL_POR_CATEGORIA: Readonly<Record<string, string>> = {
  FORMULARIO: "RESTRITO",
  FICHA_EPI: "RESTRITO",
  LISTA_PRESENCA: "RESTRITO",
  CERTIFICADO: "RESTRITO",
  CAT_SP: "RESTRITO",
  RELATORIO_INVESTIGACAO: "RESTRITO",
  EVIDENCIA_ACIDENTE: "RESTRITO",
};

/**
 * O teto do arquivo na nuvem: 4 MB. Abaixo do corpo máximo da função do
 * Netlify (6 MB, ≈4,5 MB de binário depois da codificação) com folga para os
 * outros campos do formulário.
 */
export const LIMITE_BYTES = 4 * 1024 * 1024;

export class AnexoGrandeDemais extends RegraViolada {
  constructor(public tamanho: number) {
    super(
      `O arquivo tem ${(tamanho / 1024 / 1024).toFixed(1).replace(".", ",")} MB e o limite é ` +
        `${LIMITE_BYTES / 1024 / 1024} MB por arquivo nesta instalação. ` +
        "Reduza o PDF (\"Salvar como PDF reduzido\" ou imprimir em PDF com qualidade menor) " +
        "ou divida-o em partes, e envie de novo.",
    );
  }
}

export interface ResultadoAnexo {
  anexo: Anexo;
  duplicado: boolean;
  tambem_em: number;
}

export function digerir(conteudo: Uint8Array): string {
  return createHash("sha256").update(conteudo).digest("hex");
}

/** O `storage_key` do blob — o mesmo formato do Python (`ab/abcdef...`). */
export function chave_do_blob(sha256: string): string {
  if (!/^[0-9a-f]{64}$/.test(sha256)) throw new Error(`sha256 invalido: ${sha256}`);
  return `${sha256.slice(0, 2)}/${sha256}`;
}

/** A chave no armazenamento: o `storage_key` sob o prefixo `anexos/`. */
export function chave_no_armazenamento(storage_key: string): string {
  return `anexos/${storage_key}`;
}

/** `Path(nome).stem` e `.suffix` do Python. */
function haste_e_sufixo(nome: string): [string, string] {
  const base = nome.split("/").pop() ?? nome;
  const i = base.lastIndexOf(".");
  if (i > 0 && i < base.length - 1) return [base.slice(0, i), base.slice(i)];
  return [base, ""];
}

export async function guardar(
  tx: Executor,
  d: {
    entidade: string;
    entidade_id: number;
    nome_original: string;
    conteudo: Uint8Array;
    mime_type: string;
    categoria: string;
    usuario: UsuarioAtual;
    processo_id?: number | null;
    numero_documento_sei?: string | null;
    assinado?: boolean;
    origem_migracao?: string | null;
    origem_ref?: string | null;
  },
): Promise<ResultadoAnexo> {
  if (!CATEGORIAS.includes(d.categoria)) throw new RegraViolada(`categoria inválida: ${d.categoria}`);
  if (d.conteudo.byteLength > LIMITE_BYTES) throw new AnexoGrandeDemais(d.conteudo.byteLength);

  const sha256 = digerir(d.conteudo);

  const [existente] = await tx
    .select()
    .from(tabela_anexo)
    .where(
      and(
        eq(tabela_anexo.entidade, d.entidade),
        eq(tabela_anexo.entidade_id, d.entidade_id),
        eq(tabela_anexo.sha256, sha256),
      ),
    );
  if (existente) {
    await auditoria.registrar(tx, {
      entidade: "anexo",
      entidade_id: existente.id,
      processo_id: d.processo_id ?? null,
      tipo_evento: auditoria.ANEXO_DUPLICADO,
      descricao:
        `Arquivo '${d.nome_original}' ignorado: já anexado como ` +
        `'${existente.nome_original}' (mesmo SHA-256).`,
      usuario: d.usuario,
    });
    return { anexo: existente, duplicado: true, tambem_em: await quantos_processos(tx, sha256) };
  }

  if (d.categoria === "PARECER_ASSINADO") {
    await tx
      .update(tabela_anexo)
      .set({ ativo: false })
      .where(
        and(
          eq(tabela_anexo.entidade, d.entidade),
          eq(tabela_anexo.entidade_id, d.entidade_id),
          eq(tabela_anexo.categoria, "PARECER_ASSINADO"),
          eq(tabela_anexo.ativo, true),
        ),
      );
  }

  const storage_key = chave_do_blob(sha256);
  const armazenamento = obterArmazenamento();
  const chave = chave_no_armazenamento(storage_key);
  if (!(await armazenamento.existe(chave))) {
    await armazenamento.gravar(chave, d.conteudo, d.mime_type || "application/octet-stream");
  }

  const [haste, sufixo] = haste_e_sufixo(d.nome_original);
  const [anexo] = await tx
    .insert(tabela_anexo)
    .values({
      entidade: d.entidade,
      entidade_id: d.entidade_id,
      nome_arquivo: [...slug_ascii(haste)].slice(0, 200).join("") + sufixo.toLowerCase(),
      nome_original: d.nome_original,
      mime_type: d.mime_type || "application/octet-stream",
      tamanho_bytes: d.conteudo.byteLength,
      sha256,
      storage_key,
      categoria: d.categoria,
      nivel_acesso: NIVEL_POR_CATEGORIA[d.categoria] ?? "PUBLICO",
      numero_documento_sei: d.numero_documento_sei ?? null,
      assinado: d.assinado ?? false,
      origem_migracao: d.origem_migracao ?? null,
      origem_ref: d.origem_ref ?? null,
      enviado_por: d.usuario.id,
    })
    .returning();

  await auditoria.registrar(tx, {
    entidade: "anexo",
    entidade_id: anexo!.id,
    processo_id: d.processo_id ?? null,
    tipo_evento: "ANEXO_ENVIADO",
    descricao: `'${d.nome_original}' (${d.categoria}, ${d.conteudo.byteLength} bytes).`,
    usuario: d.usuario,
  });
  return { anexo: anexo!, duplicado: false, tambem_em: await quantos_processos(tx, sha256) };
}

/** O `guardar_arquivo` do Python, a partir do `File` do formulário. */
export async function guardar_arquivo(
  tx: Executor,
  arquivo: File,
  d: Omit<Parameters<typeof guardar>[1], "conteudo" | "nome_original" | "mime_type"> & { mime_type?: string },
): Promise<ResultadoAnexo> {
  // recusa antes de ler: o `File` já está na memória, mas não precisa virar
  // uma segunda cópia para ser recusado
  if (arquivo.size > LIMITE_BYTES) throw new AnexoGrandeDemais(arquivo.size);
  return guardar(tx, {
    ...d,
    conteudo: new Uint8Array(await arquivo.arrayBuffer()),
    nome_original: arquivo.name,
    mime_type: d.mime_type ?? arquivo.type,
  });
}

async function quantos_processos(tx: Executor, sha256: string): Promise<number> {
  const linhas = await tx
    .select({ entidade: tabela_anexo.entidade, entidade_id: tabela_anexo.entidade_id })
    .from(tabela_anexo)
    .where(eq(tabela_anexo.sha256, sha256));
  return new Set(linhas.map((l) => `${l.entidade}\u0000${l.entidade_id}`)).size;
}

export async function listar(tx: Executor, entidade: string, entidade_id: number): Promise<Anexo[]> {
  return tx
    .select()
    .from(tabela_anexo)
    .where(
      and(eq(tabela_anexo.entidade, entidade), eq(tabela_anexo.entidade_id, entidade_id), eq(tabela_anexo.ativo, true)),
    )
    .orderBy(desc(tabela_anexo.enviado_em));
}

/** O conteúdo do anexo (o `caminho_absoluto` + `read_bytes` do Python). */
export async function ler_conteudo(anexo: Pick<Anexo, "storage_key">): Promise<Uint8Array> {
  return obterArmazenamento().ler(chave_no_armazenamento(anexo.storage_key));
}

/** O conteúdo do blob pelo SHA-256 (a rubrica do certificado). */
export async function ler_blob(sha256: string): Promise<Uint8Array> {
  return obterArmazenamento().ler(chave_no_armazenamento(chave_do_blob(sha256)));
}

export async function blob_existe(sha256: string): Promise<boolean> {
  return obterArmazenamento().existe(chave_no_armazenamento(chave_do_blob(sha256)));
}
